### PR TITLE
Added a warning message while accessing credentials through rails command

### LINKED ISSUE #41830: Use RAILS_ENV to edit different credentials files
### Steps to reproduce

1. Run `rails credentials:edit` in development environment and add/edit credentials
2. Run `RAILS_ENV=production rails credentials:edit` to edit production credentials but same file opens up

### Expected behaviour
I expected `RAILS_ENV=production rails credentials:edit` to work much like `RAILS_ENV=production rails console`

### Actual behaviour
You actually need to specify the `--environment` flag in order to change environments. This is documented in the [PR introducing the feature](https://github.com/rails/rails/pull/33521) but not in the [Rails Guide](https://edgeguides.rubyonrails.org/security.html).

It also means you can run confusing commands like `RAILS_ENV=staging rails credentials:edit --environment production`

I suspect it was done this way to maintain backwards compatibility. It's only a minor paper cut but to me it violates the principle of least surprise.

If it needs to remain this way for backwards compatibility then clearer documentation would help (I'm happy to do a PR if that's the way forward). Additionally, if there is any way to detect someone has explicitly passed `RAILS_ENV` to this command, a warning explaining the correct way to edit the environment along with a link to the documentation would be very helpful.

### System configuration
**Rails version**: Rails 6.1.3

**Ruby version**: ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-darwin20]

### PR DESCRIPTION
**Pull request by Navdevl**
### Summary

Fixes #41830 

The current implementation shows an non-blocking warning message to the user when running `bin/rails credentials:show` or `bin/rails credentials:edit` from an non-development environment and used `--environment` different from the current environment of the system. 

That is, when the `credentials` commands are ran from a production environment but with `--environment=staging` flag, the user will be shown a warning message. 

### Other Information

From the discussion at #41830, I also thought if we could add to show_command also, it would be beneficial. This is my first time contribution to the Rails code (One earlier contribution was rejected, so not counting that in 😅), so kindly help me out if the code I introduced isn't as per the Rails convention or I should be thinking from different perspective or maybe add more thoughts with certain implementation areas.

### CODE REVIEW

File: guides/source/security.md
Comment (ghiculescu) [inline guides/source/security.md:1220]: I don't think this part is necessary. Let's show the user the correct way to do it, they will see the warning if they get it wrong.

Comment (Navdevl) [reply guides/source/security.md:1220]: Uhm, @ghiculescu mind sharing a little more context please? So which part do you think is not necessary. I mean, I am currently showing the warning message when the `rails credentials:edit` or `rails credentials:show` is ran under certain scenario. I am little bit confused about the part that is not necessary. 😅

Comment (ghiculescu) [reply guides/source/security.md:1220]: (I'm not a core committer so ultimately it's not up to me, but) I don't think lines 1218 - 1220 are necessary. I think it's better to keep the docs concise and tell you what you need to know to do the task correctly.

Comment (Navdevl) [reply guides/source/security.md:1220]: Oh, got it. I'll reword the added docs to be more concise. Thanks!! 😄

Comment (Navdevl) [reply guides/source/security.md]: @ghiculescu Made the docs more concise. Now it delivers only the needed information.

File: guides/source/security.md
Comment (zzak) [inline guides/source/security.md:1213]:
```suggestion
For example, to edit a production credential file, you need to run:

```

Comment (zzak) [reply guides/source/security.md]: @navdevl Could you also add the newline between the block? :bow:

Comment (Navdevl) [reply guides/source/security.md]: @zzak thanks for the review comments. I have made the changes now. 😄

Comment (Navdevl) [reply guides/source/security.md]: @zzak Just asking out of curiosity. I am pushing on the same commit by squashing it. Is that how it should be? Or I should be adding a new commit for every review happening and squash them at the last? How things are done usually? 😅😅 _(I know the question sounds silly)_

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (zzak) [inline railties/lib/rails/commands/credentials/credentials_command.rb:124]: I think this method is simple enough that this comment is almost unnecessary, since it's private only Rails devs will see it.

If we do want to add a comment, I would try to rephrase it to be more concise, for example:

> Check for a mismatch between Rails::Command.environment and option set by --environment flag. This warning should only occur in non-development environments to avoid confusing the user when trying to perform this action against different environment modes.

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (zzak) [inline railties/lib/rails/commands/credentials/credentials_command.rb:127]: Knit: but could we line-break this string?

Something like `"you are accessing....." \ "blah blah" \ "foo bar"?

File: guides/source/security.md
Comment (zzak) [inline guides/source/security.md:1209]:
```suggestion
#### Accessing environment specific credentials
```

Comment (Navdevl) [reply guides/source/security.md]: Updated the heading. 😄

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (Navdevl) [reply railties/lib/rails/commands/credentials/credentials_command.rb]: @zzak Introduced line breaks to make it little more clear to the user. Thanks for the suggestion.

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (Navdevl) [reply railties/lib/rails/commands/credentials/credentials_command.rb]: Updated the comments.

File: guides/source/security.md
Comment (zzak) [reply guides/source/security.md]: @Navdevl Squashing to a single commit is preferred, so what you're doing is exactly perfect :)

But don't worry about doing it everytime, sometimes it's better to wait for a committer or core to review before squashing since they may have more changes.

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (zzak) [inline railties/lib/rails/commands/credentials/credentials_command.rb:128]:
```suggestion
              For more information try `rails credentials:help`.
```

File: guides/source/security.md
Comment (Navdevl) [reply guides/source/security.md]: Oh, that makes sense. Understood. Thanks @zzak! :))

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (Navdevl) [reply railties/lib/rails/commands/credentials/credentials_command.rb]: @zzak Updated.

**Comment by zzak** [comment]
@Navdevl Could you squash please? :bow:

**Comment by Navdevl** [comment]
@zzak Squashed into single commit. :)

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (sunny) [inline railties/lib/rails/commands/credentials/credentials_command.rb:128]: Prefer using `bin/rails` here, as it’ll be copy-pasteable and work everywhere:

```suggestion
              Did you mean to run `bin/rails credentials:#{method_name} --environment #{Rails::Command.environment}`?
              For more information try `bin/rails credentials:help`.
```

File: railties/test/commands/credentials_test.rb
Comment (sunny) [inline railties/test/commands/credentials_test.rb:95]: (If applying previous suggestion.)

```suggestion
      assert_match(/You are accessing production credentials from the staging environment\.\nDid you mean to run `bin/rails credentials:edit --environment staging`\?\nFor more information try `bin/rails credentials:help`/, output)
```

File: railties/test/commands/credentials_test.rb
Comment (sunny) [inline railties/test/commands/credentials_test.rb:134]: (If applying previous suggestion.)

```suggestion
      assert_match(/You are accessing production credentials from the staging environment\.\nDid you mean to run `bin/rails credentials:show --environment staging`\?\nFor more information try `bin/rails credentials:help`/, output)
```

File: guides/source/security.md
Comment (Navdevl) [reply guides/source/security.md]: Wow. I missed a whole lot of things to look for. Thanks @sunny for pointing that out. Addressed the comments in the latest commit. :))

File: guides/source/security.md
Comment (Navdevl) [reply guides/source/security.md:1213]: Makes sense.

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (Navdevl) [reply railties/lib/rails/commands/credentials/credentials_command.rb]: Noted. I'll follow the same from now on. :))

**Comment by Navdevl** [comment]
@sunny @zzak Should I now squash my commits to one commit?

**Comment by Navdevl** [comment]
@sunny @zzak I squashed the commits to a single commit now.

**Comment by Navdevl** [comment]
@p8 friendly reminder on this! :))

**Review by zzak — APPROVED** [review]

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (p8) [inline railties/lib/rails/commands/credentials/credentials_command.rb:124]: I think we generally try to avoid calling environments inside Rails and try to use a configuration instead. See: https://github.com/rails/rails/pull/42828#issuecomment-884470990

**Review by p8 — CHANGES_REQUESTED** [review]

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (Navdevl) [reply railties/lib/rails/commands/credentials/credentials_command.rb:124]: @p8 Sorry to be bothering with simple questions. So I went through the mentioned PR and understood that we can use values from `request.env` in certain places. In our case, it is a command line execution right? So, ideally that can't be possible from request object.

So should I be looking in for a different variable that holds the configuration?

**Review by marivaldo — APPROVED** [review]

File: railties/lib/rails/commands/credentials/credentials_command.rb
Comment (p8) [reply railties/lib/rails/commands/credentials/credentials_command.rb:124]: No problem 🙂 You can ignore the request.env stuff. I think we should add a config option, maybe config.credentials.allow_editing, that is only set to true for the development environment. 
Similar to: https://github.com/rails/rails/blob/52db7f2ef3c4a902227cac163657c89282986e25/railties/lib/rails/application/configuration.rb#L37 and https://github.com/rails/rails/blob/52db7f2ef3c4a902227cac163657c89282986e25/railties/lib/rails/application/configuration.rb#L72

**Comment by Navdevl** [comment]
Sorry was busy for a while. I'll be picking this up and working on the comments in the next couple days.

**Pull request closed without merging**
